example-dvc-experiments/code/src/train.py

```diff
@@ -16,6 +18,93 @@
 BATCH_SIZE = 128
 
 
+def label_from_path(filepath):
+    """extracts "test", and 3 from a path like "images/test/3/00177.png" """
+    elements = filepath.split('/')
+    return (elements[1], int(elements[2]))
+
+def read_dataset(dataset_path):
+    ds = tarfile.open(name=dataset_path, mode='r:gz')
+    training, testing = [], []
+    print(f"Reading dataset from {dataset_path}")
+    for f in ds:
+        if f.isfile():
+            filepath = f.name
+            content = ds.extractfile(f)
+            image = imageio.imread(content)
+            imagesection, imagelabel = label_from_path(filepath)
+            if imagesection == "train":
+                training.append((imagelabel, image))
+            else:
+                testing.append((imagelabel, image))
+    training_len = len(training)
+    testing_len = len(testing)
+    print(f"Read {training_len} training images and {testing_len} testing images")
+    # we assume the images are 28x28 grayscale
+    shape_0, shape_1 = 28, 28
+    testing_images = np.ndarray(shape=(len(testing), shape_0, shape_1), dtype="uint8")
+    testing_labels = np.zeros(shape=(len(testing)), dtype="uint8")
+    for i, (label, image) in enumerate(testing):
+        testing_images[i] = image
+        testing_labels[i] = label
+    training_images = np.ndarray(shape=(len(training), shape_0, shape_1), dtype="uint8")
+    training_labels = np.zeros(shape=(len(training)), dtype="uint8")
+    for i, (label, image) in enumerate(training):
+        training_images[i] = image
+        training_labels[i] = label
+    return (training_images, training_labels, testing_images, testing_labels)
+
+
+def create_image_matrix(cells):
```

Comment:
While this is a cool visual, it does add a quite a bit to the example code. What about using some built-in visualization, like https://keras.io/api/utils/model_plotting_utils/#plot_model-function? It's probably much less useful, but it's certainly less example code. Not a strong opinion if others prefer this visual.

Reply:
IMO, our model structure is a bit static and not-so-exciting. It's very simple to visualize THB 😅 

I can think of other, simpler ways. This came to my mind as a "confusion matrix" in image form. (I was thinking to put all classification errors in tiny little boxes on a large image, but one from each "confusion" looked better.)

The code is not that visible to the users. I don't think they'll peek inside the code that much. We'll just show the results and (maybe) link to the code in Github.

Reply:
Moved this code to `util.py`. I think we can consider this resolved.